arch/xtensa/src/esp32s3/esp32s3_spi_timing.c

```diff
@@ -238,129 +349,1232 @@ static void IRAM_ATTR set_flash_clock(uint8_t spi_num, uint32_t freqdiv)
+ *   din_mode - Din mode
+ *   din_num  - Din num
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+static void set_psram_din_mode_num(uint8_t spi_num, uint8_t din_mode,
+                                   uint8_t din_num)
+{
+  /* Set MSPI PSRAM din mode */
+
+  uint32_t reg_val = (REG_READ(SPI_MEM_SPI_SMEM_DIN_MODE_REG(spi_num)) &
+      (~(SPI_MEM_SPI_SMEM_DIN0_MODE_M | SPI_MEM_SPI_SMEM_DIN1_MODE_M |
+         SPI_MEM_SPI_SMEM_DIN2_MODE_M | SPI_MEM_SPI_SMEM_DIN3_MODE_M |
+         SPI_MEM_SPI_SMEM_DIN4_MODE_M | SPI_MEM_SPI_SMEM_DIN5_MODE_M |
+         SPI_MEM_SPI_SMEM_DIN6_MODE_M | SPI_MEM_SPI_SMEM_DIN7_MODE_M |
+         SPI_MEM_SPI_SMEM_DINS_MODE_M))) |
+         (din_mode << SPI_MEM_SPI_SMEM_DIN0_MODE_S) |
+         (din_mode << SPI_MEM_SPI_SMEM_DIN1_MODE_S) |
+         (din_mode << SPI_MEM_SPI_SMEM_DIN2_MODE_S) |
+         (din_mode << SPI_MEM_SPI_SMEM_DIN3_MODE_S) |
+         (din_mode << SPI_MEM_SPI_SMEM_DIN4_MODE_S) |
+         (din_mode << SPI_MEM_SPI_SMEM_DIN5_MODE_S) |
+         (din_mode << SPI_MEM_SPI_SMEM_DIN6_MODE_S) |
+         (din_mode << SPI_MEM_SPI_SMEM_DIN7_MODE_S) |
+         (din_mode << SPI_MEM_SPI_SMEM_DINS_MODE_S);
+  REG_WRITE(SPI_MEM_SPI_SMEM_DIN_MODE_REG(spi_num), reg_val);
+
+  /* Set MSPI PSRAM din num */
+
+  reg_val = (REG_READ(SPI_MEM_SPI_SMEM_DIN_NUM_REG(spi_num)) &
+      (~(SPI_MEM_SPI_SMEM_DIN0_NUM_M | SPI_MEM_SPI_SMEM_DIN1_NUM_M |
+         SPI_MEM_SPI_SMEM_DIN2_NUM_M | SPI_MEM_SPI_SMEM_DIN3_NUM_M |
+         SPI_MEM_SPI_SMEM_DIN4_NUM_M | SPI_MEM_SPI_SMEM_DIN5_NUM_M |
+         SPI_MEM_SPI_SMEM_DIN6_NUM_M | SPI_MEM_SPI_SMEM_DIN7_NUM_M |
+         SPI_MEM_SPI_SMEM_DINS_NUM_M))) |
+         (din_num << SPI_MEM_SPI_SMEM_DIN0_NUM_S) |
+         (din_num << SPI_MEM_SPI_SMEM_DIN1_NUM_S) |
+         (din_num << SPI_MEM_SPI_SMEM_DIN2_NUM_S) |
+         (din_num << SPI_MEM_SPI_SMEM_DIN3_NUM_S) |
+         (din_num << SPI_MEM_SPI_SMEM_DIN4_NUM_S) |
+         (din_num << SPI_MEM_SPI_SMEM_DIN5_NUM_S) |
+         (din_num << SPI_MEM_SPI_SMEM_DIN6_NUM_S) |
+         (din_num << SPI_MEM_SPI_SMEM_DIN7_NUM_S) |
+         (din_num << SPI_MEM_SPI_SMEM_DINS_NUM_S);
+  REG_WRITE(SPI_MEM_SPI_SMEM_DIN_NUM_REG(spi_num), reg_val);
+}
+
+#if ESP32S3_SPI_TIMING_FLASH_TUNING
+
+/****************************************************************************
+ * Name: config_flash_read_data
+ *
+ * Description:
+ *   Configure Flash to read data via SPI1h
+ *
+ * Input Parameters:
+ *   buf  - data buffer pointer
+ *   addr - target address value
+ *   len  - data length
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+static void config_flash_read_data(uint8_t *buf, uint32_t addr, uint32_t len)
+{
+#ifdef CONFIG_ESP32S3_FLASH_MODE_OCT
+  /* Clear MSPI hw fifo */
+
+  for (int i = 0; i < 16; i++)
     {
-      set_flash_clock(1, flash_div);
+      REG_WRITE(SPI_MEM_W0_REG(1) + i * 4, 0);
```

Comment:
Why not putreg32?

Reply:
Already replaced, REG_WRITE is the same as putreg32.